src/libstd/thread/local.rs

```diff
@@ -364,54 +398,84 @@ pub mod fast {
+        // performed for every call to `Key::get`.
+        // LLVM issue: https://bugs.llvm.org/show_bug.cgi?id=41722
+        #[inline(never)]
+        #[cold]
+        unsafe fn try_initialize_nodrop<F: FnOnce() -> T>(&self, init: F) -> &'static T {
+            self.inner.initialize(init)
+        }
+
+        // `try_initialize_drop` is only called once per fast thread local
+        // variable, except in corner cases where thread_local dtors reference
+        // other thread_local's, or it is being recursively initialized.
+        #[inline(never)]
+        #[cold]
+        unsafe fn try_initialize_drop<F: FnOnce() -> T>(&self, init: F) -> Option<&'static T> {
+            // We don't put a `needs_drop` check around this and call it a day
+            // because this function is not inlined. Unwrapping code gets
+            // generated for callers of `LocalKey::with` even if we always
+            // return `Some` here.
+            match self.dtor_state.get() {
+                DtorState::Unregistered => {
+                    // dtor registration happens before initialization.
+                    register_dtor(self as *const _ as *mut u8,
+                                destroy_value::<T>);
+                    self.dtor_state.set(DtorState::Registered);
+                }
+                DtorState::Registered => {
+                    // recursively initialized
+                }
+                DtorState::RunningOrHasRun => {
+                    return None
+                }
             }
 
-            register_dtor(self as *const _ as *mut u8,
-                          destroy_value::<T>);
-            self.dtor_registered.set(true);
+            Some(self.inner.initialize(init))
         }
     }
 
     unsafe extern fn destroy_value<T>(ptr: *mut u8) {
         let ptr = ptr as *mut Key<T>;
-        // Right before we run the user destructor be sure to flag the
-        // destructor as running for this thread so calls to `get` will return
-        // `None`.
-        (*ptr).dtor_running.set(true);
 
-        // Some implementations may require us to move the value before we drop
-        // it as it could get re-initialized in-place during destruction.
-        //
-        // Hence, we use `ptr::read` on those platforms (to move to a "safe"
-        // location) instead of drop_in_place.
-        if requires_move_before_drop() {
-            ptr::read((*ptr).inner.get());
-        } else {
-            ptr::drop_in_place((*ptr).inner.get());
-        }
+        // Right before we run the user destructor be sure to set the
+        // `Option<T>` to `None`, and `dtor_state` to `RunningOrHasRun`. This
+        // causes future calls to `get` to run `try_initialize_drop` again,
+        // which will now fail, and return `None`.
+        let value = (*ptr).inner.take();
+        (*ptr).dtor_state.set(DtorState::RunningOrHasRun);
+        drop(value);
```

Comment:
This is technically subtly different from before where it's moving the value out and then dropping it (as opposed to dropping it in place), but I wanted to confirm to see if that was intentional?)

Reply:
This was intentional. We need to set the option to `None` before dropping. This guards against the scenario where the destructor attempts to reaccess the `thread_local!`. It's a bit unfortunate, but up until a few months ago macos did a similar thing.

Reply:
Hm are you sure? I was under the impression that `RunningOrHasRun` would prevent all future accesses which means we're allowed to drop in place (like the previous implementation). The OSX quirk was historically because it was unsafe to drop-in-place becase it would also sometimes reinitialize in place while we're destroying in place

Reply:
Sorry for the late reply!

In order to get the single `cmp je` fast path (and the single `tlv_get_addr` call on OSX) the `UnsafeCell<Option<T>>` type is used to either mean:
```rust
Some(reference) => { /* access away! no problems */ }
None => { /* check dtor_state for more information */ }
```
In other words, `dtor_state` is never examined unless the `Option` is set to `None`.

Reply:
Aha I missed that subtelty! I think I should have read the docs more closely, thanks for the explainer!